shared/shared.go

```diff
@@ -45,3 +48,20 @@ type MerkleProof struct {
 	ProvenLeaves [][]byte
 	ProofNodes   [][]byte
 }
+
+func Retry(retryable func() error, numRetries int, interval time.Duration, logger func(msg string)) error {
+	err := retryable()
+	if err == nil {
+		return nil
+	}
+
+	if numRetries < 1 {
+		logger(err.Error())
+		return errors.New("number of retries exceeded")
```

Comment:
I prefer to print a single message, so instead of printing the error and then returning a static message, I'd return:
```golang
return fmt.Errorf("number of retries exceeded: %v", err)
```

Reply:
Errors returned from the function calls are logged and not aggregated, to report them immediately. The returned error is only about the overall retries exhaustion. This is why I didn't combine it with the returned error as you suggested.